### ISSUE TITLE
Add Log Cosh Loss

### ISSUE THREAD

**Issue by kanavgoyal898** [pytorch/pytorch#190037]
# Feature Request: Add `torch.nn.LogCoshLoss`

## Summary

I would like to work on adding a native implementation of **Log-Cosh Loss** to PyTorch and would appreciate feedback from the maintainers on the proposed API and implementation approach before opening a pull request.

**If this proposal is accepted, I would like to work on this feature and submit the implementation.**

## Motivation

`LogCoshLoss` is a smooth, twice-differentiable alternative to Mean Squared Error (MSE) and Mean Absolute Error (MAE). It behaves like:

-   **MSE** for small errors, providing smooth gradients.
-   **MAE** for large errors, making it more robust to outliers.

It is commonly used in regression tasks where robustness and stable optimization are both desirable.

Currently, users must implement this loss manually in Python despite it being a frequently reused objective.

## Mathematical Definition

For prediction `x` and target `y`,

``` text
loss(x, y) = log(cosh(x - y))
```

The reduction modes would follow existing PyTorch loss functions:

-   `"none"`
-   `"mean"`
-   `"sum"`

## Proposed API

Functional API:

``` python
torch.nn.functional.log_cosh_loss(
    input,
    target,
    reduction="mean",
)
```

Module API:

``` python
criterion = torch.nn.LogCoshLoss(reduction="mean")
loss = criterion(prediction, target)
```

This mirrors the design of existing losses such as `MSELoss`, `L1Loss`, `SmoothL1Loss`, and `HuberLoss`.

## Implementation

The implementation would include:

-   `torch.nn.functional.log_cosh_loss`
-   `torch.nn.LogCoshLoss`
-   Documentation
-   Unit tests covering:
    -   reduction modes
    -   gradient correctness
    -   broadcasting
    -   dtype compatibility
    -   CPU and CUDA (where applicable)

## Numerical Stability

A numerically stable implementation should be used instead of the naive
expression:

``` python
torch.log(torch.cosh(x))
```

to avoid overflow for large values.

## Prior Art

Several machine learning libraries and user implementations provide Log-Cosh Loss because of its usefulness for regression and robust optimization. A native implementation in PyTorch would improve discoverability, consistency, and usability.

## References

-   TensorFlow Log-Cosh Loss:
    https://www.tensorflow.org/api_docs/python/tf/keras/losses/LogCosh

## Additional Notes

I have reviewed the existing loss function APIs in `torch.nn` and intend to follow the current implementation patterns, documentation style, and testing conventions used throughout the project.

I would appreciate any feedback on the API or implementation approach before starting the work.

Thank you!

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by albanD** [comment]
Hey @kanavgoyal898 , thanks for the detailed request! Note that we maintain a very high bar for 
inclusion of new modules within torch.nn, as each comes with a substantial maintenance cost on our 
end. In general, we will accept new modules if the underlying techniques have already achieved 
widespread adoption and there is a broad expectation that PyTorch will provide such a module. It's 
also beneficial if there are performance reasons why the module should be provided by PyTorch itself 
rather than in a third-party repo.
Details about this can be found in https://github.com/pytorch/pytorch/wiki/Developer-FAQ#i-have-a-new-function-or-feature-id-like-to-add-to-pytorch-should-it-be-in-pytorch-core-or-a-library-like-torchvision

The numerical stability argument sounds good, but I'm curious how broadly used it is.

**Comment by kanavgoyal898** [comment]
Thanks for the feedback!

That makes sense. I understand the high bar for adding new `torch.nn` modules.

My reasoning for proposing `LogCoshLoss` is that it is already a fairly common regression loss, particularly in robust regression settings, and is exposed as a first-class loss in TensorFlow/Keras. There are also many independent PyTorch implementations across tutorials, repositories, and libraries, which suggests users repeatedly reimplement the same functionality rather than relying on a standard implementation.

Beyond convenience, I think there are two potential benefits of including it in core:

* A numerically stable implementation can be provided by default, avoiding the overflow issues of the naive `log(cosh(x))` formulation.
* It would provide a consistent API alongside existing regression losses such as `MSELoss`, `L1Loss`, `SmoothL1Loss`, and `HuberLoss`, with full support for reductions, autograd, broadcasting, devices, and dtypes.

That said, I'll gather more concrete evidence of adoption (e.g., usage in popular libraries, research, and community implementations) to better justify that it meets the bar for inclusion before proceeding further.

Thanks again for the guidance!

**Comment by kanavgoyal898** [comment]
Some evidence on adoption:

- **TensorFlow/Keras provides Log-Cosh as a first-class built-in loss**, available through both the functional and module APIs:
  - https://www.tensorflow.org/api_docs/python/tf/keras/losses/LogCosh
  - https://keras.io/api/losses/regression_losses/#logcosh-class

- **The statistical properties of Log-Cosh have been studied in dedicated ML research**, highlighting its use as a robust regression loss:
  - Saleh & Saleh, *Statistical Properties of the log-cosh Loss Function Used in Machine Learning* (2022): https://arxiv.org/abs/2208.04564

- **Recent surveys of regression loss functions include Log-Cosh alongside standard objectives such as MSE, MAE, Huber, and SmoothL1**, indicating that it is recognized as a standard regression loss:
  - Jadon et al., *A Comprehensive Survey of Regression Based Loss Functions for Time Series Forecasting* (2022): https://arxiv.org/abs/2211.02989
  - *A Comprehensive Survey of Loss Functions and Metrics in Deep Learning* (Artificial Intelligence Review, 2025): https://link.springer.com/article/10.1007/s10462-025-11198-7

cc @albanD
